server/resp/src/main/java/org/infinispan/server/resp/commands/sortedset/ZRANGE.java

```diff
@@ -3,49 +3,265 @@
+         getSortedSetCall = sortedSetCache.subsetByScore(name, builder.build());
+
+      } else if (byLex) {
+         // ZRANGE by lex. Replaces ZRANGEBYLEX and ZREVRANGEBYLEX
+         SortedSetArgumentsUtils.Lex startLex = SortedSetArgumentsUtils.parseLex(start);
+         SortedSetArgumentsUtils.Lex stopLex = SortedSetArgumentsUtils.parseLex(stop);
+         if (startLex == null || stopLex == null) {
+            RespErrorUtil.customError("min or max not valid string range item", handler.allocator());
+            return handler.myStage();
+         }
+         if ((isRev && startLex.unboundedMin) || (!isRev && startLex.unboundedMax)
+               || (startLex.unboundedMin && stopLex.unboundedMin)
+               || (startLex.unboundedMax && stopLex.unboundedMax)) {
+            // no call when + in min or rev with - is empty, or when we get + + or - -
+            return returnEmptyList(handler, ctx, name, sortedSetCache);
+         }
+
+         SortedSetSubsetArgs.Builder<byte[]> builder = SortedSetSubsetArgs.create();
+         builder.isRev(isRev);
+         builder.start(startLex.value);
+         builder.includeStart(startLex.include);
+         builder.stop(stopLex.value);
+         builder.includeStop(stopLex.include);
+         getSortedSetCall = sortedSetCache.subsetByLex(name, builder.build());
+      } else {
+         // ZRANGE by index. Replaces ZREVRANGE
+         long from;
+         long to;
+         try {
+            from = ArgumentUtils.toLong(start);
+            to = ArgumentUtils.toLong(stop);
+         } catch (NumberFormatException ex) {
+            RespErrorUtil.valueNotInteger(handler.allocator());
+            return handler.myStage();
+         }
+         SortedSetSubsetArgs.Builder<Long> builder = SortedSetSubsetArgs.create();
+         builder.isRev(isRev);
+         builder.start(from);
+         builder.stop(to);
+         getSortedSetCall = sortedSetCache.subsetByIndex(name, builder.build());
+      }
+      CompletionStage<List<byte[]>> list = getSortedSetCall.thenApply(subsetResult -> mapResultsToArrayList(subsetResult, resultOpt));
       return handler.stageToReturn(list, ctx, Consumers.GET_ARRAY_BICONSUMER);
    }
+
+   private static CompletionStage<RespRequestHandler> returnEmptyList(Resp3Handler handler,
+                                                                      ChannelHandlerContext ctx,
+                                                                      byte[] name,
+                                                                      EmbeddedMultimapSortedSetCache<byte[], byte[]> sortedSetCache) {
+      // Size is called to raise ERR type if needed
+      return handler.stageToReturn(sortedSetCache.size(name)
+                  .thenApply(r -> Collections.emptyList()), ctx, Consumers.GET_ARRAY_BICONSUMER);
+   }
+
+   /**
+    * Transforms the resulting collection depending on the zrank options
+    * - return scores
+    * - limit results
```

Comment:
Thinking if we send these parameters to limit the remote results. Maybe we add it later.

Reply:
yes, I've thought on that too as an improvement of perf, we need to run some tests to go into improv